Possible issue with bitshuffle filter on big-endian machines

This has been discovered by the PyTables test suite (where a C-Blosc implementing bitshuffle has been vendored into it recently): https://github.com/PyTables/PyTables/issues/583

@kiyo-masui any hints here?

Here are the result on an s390 Debian porterbox:
```
$ nosetests3 bitshuffle
FF..F...F..........F.......F........F.........F..F...F.FFFFEF
======================================================================
ERROR: test_plugins (bitshuffle.tests.test_h5plugin.TestFilterPlugins)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pini/debian/bitshuffle/bitshuffle/tests/test_h5plugin.py", line 55, in test_plugins
    stdout=PIPE, stderr=STDOUT)
  File "/usr/lib/python3.5/subprocess.py", line 947, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.5/subprocess.py", line 1551, in _execute_child
    raise child_exception_type(errno_num, err_msg)
FileNotFoundError: [Errno 2] No such file or directory: 'h5dump'

======================================================================
FAIL: test_circle (bitshuffle.tests.test_ext.TestBitShuffleCircle)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pini/debian/bitshuffle/bitshuffle/tests/test_ext.py", line 517, in test_circle
    == out.view(np.uint8)))
AssertionError: False is not true

======================================================================
FAIL: test_circle_with_compression (bitshuffle.tests.test_ext.TestBitShuffleCircle)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pini/debian/bitshuffle/bitshuffle/tests/test_ext.py", line 534, in test_circle_with_compression
    == out.view(np.uint8)))
AssertionError: False is not true

======================================================================
FAIL: test_trans_bit_elem_scal (bitshuffle.tests.test_ext.TestOddLengths)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pini/debian/bitshuffle/bitshuffle/tests/test_ext.py", line 486, in tearDown
    self.assertTrue(np.all(out == ans))
AssertionError: False is not true

======================================================================
FAIL: test_untrans_bit_elem_scal (bitshuffle.tests.test_ext.TestOddLengths)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pini/debian/bitshuffle/bitshuffle/tests/test_ext.py", line 486, in tearDown
    self.assertTrue(np.all(out == ans))
AssertionError: False is not true

======================================================================
FAIL: test_03a_trans_bit_byte (bitshuffle.tests.test_ext.TestProfile)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pini/debian/bitshuffle/bitshuffle/tests/test_ext.py", line 70, in tearDown
    self.assertTrue(np.all(ans == out.view(np.uint8)))
AssertionError: False is not true

======================================================================
FAIL: test_04e_trans_bit_elem_64 (bitshuffle.tests.test_ext.TestProfile)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pini/debian/bitshuffle/bitshuffle/tests/test_ext.py", line 70, in tearDown
    self.assertTrue(np.all(ans == out.view(np.uint8)))
AssertionError: False is not true

======================================================================
FAIL: test_06g_untrans_bit_elem_64 (bitshuffle.tests.test_ext.TestProfile)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pini/debian/bitshuffle/bitshuffle/tests/test_ext.py", line 73, in tearDown
    self.assertTrue(np.all(ans == out.view(np.uint8)))
AssertionError: False is not true

======================================================================
FAIL: test_09a_trans_bit_elem_scal_64 (bitshuffle.tests.test_ext.TestProfile)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pini/debian/bitshuffle/bitshuffle/tests/test_ext.py", line 70, in tearDown
    self.assertTrue(np.all(ans == out.view(np.uint8)))
AssertionError: False is not true

======================================================================
FAIL: test_09d_untrans_bit_elem_scal_64 (bitshuffle.tests.test_ext.TestProfile)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pini/debian/bitshuffle/bitshuffle/tests/test_ext.py", line 73, in tearDown
    self.assertTrue(np.all(ans == out.view(np.uint8)))
AssertionError: False is not true

======================================================================
FAIL: test_10b_bitunshuffle_64 (bitshuffle.tests.test_ext.TestProfile)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pini/debian/bitshuffle/bitshuffle/tests/test_ext.py", line 73, in tearDown
    self.assertTrue(np.all(ans == out.view(np.uint8)))
AssertionError: False is not true

======================================================================
FAIL: test_10d_decompress_64 (bitshuffle.tests.test_ext.TestProfile)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pini/debian/bitshuffle/bitshuffle/tests/test_ext.py", line 73, in tearDown
    self.assertTrue(np.all(ans == out.view(np.uint8)))
AssertionError: False is not true

======================================================================
FAIL: test_filter (bitshuffle.tests.test_h5filter.TestFilter)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pini/debian/bitshuffle/bitshuffle/tests/test_h5filter.py", line 37, in test_filter
    self.assertTrue(np.all(d == data))
AssertionError: False is not true

======================================================================
FAIL: test_with_block_size (bitshuffle.tests.test_h5filter.TestFilter)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pini/debian/bitshuffle/bitshuffle/tests/test_h5filter.py", line 59, in test_with_block_size
    self.assertTrue(np.all(d == data))
AssertionError: False is not true

======================================================================
FAIL: test_with_compression (bitshuffle.tests.test_h5filter.TestFilter)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pini/debian/bitshuffle/bitshuffle/tests/test_h5filter.py", line 81, in test_with_compression
    self.assertTrue(np.all(d == data))
AssertionError: False is not true

======================================================================
FAIL: test_regression (bitshuffle.tests.test_regression.TestAll)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pini/debian/bitshuffle/bitshuffle/tests/test_regression.py", line 36, in test_regression
    == g_orig[dset_name][:]))
AssertionError: False is not true

----------------------------------------------------------------------
Ran 61 tests in 0.047s

FAILED (errors=1, failures=15)
```

So I tried to compile bitshuffle on PPC64 (big endian).

1. gcc does not support `-march=native`, so I had to kick it from COMPILE_FLAGS.

2. I see the following warning:
```
src/bitshuffle_core.c: In function 'bshuf_trans_bit_byte_remainder':
src/bitshuffle_core.c:168:34: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     for (ii = start_byte / 8; ii < nbyte_bitrow; ii ++) {
                                  ^
```

3. Without your patch, the tests fail as expected (https://github.com/Blosc/c-blosc/issues/181#issuecomment-258281321)

4. With your patch it dumps core:
```
Using host libthread_db library "/lib64/power8/libthread_db.so.1".
Core was generated by `/usr/bin/python3 /usr/bin/nosetests-3.5 tests/ '.
Program terminated with signal SIGABRT, Aborted.
#0  0x00003fff7bdcf818 in __GI_raise (sig=<optimized out>) at ../sysdeps/unix/sysv/linux/raise.c:54
54	  return INLINE_SYSCALL (tgkill, 3, pid, selftid, sig);
[Current thread is 1 (Thread 0x3fff776ef190 (LWP 9266))]
(gdb) bt
#0  0x00003fff7bdcf818 in __GI_raise (sig=<optimized out>) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00003fff7bdd1f64 in __GI_abort () at abort.c:89
#2  0x00003fff7be17a28 in __libc_message (do_abort=<optimized out>, fmt=<optimized out>)
    at ../sysdeps/posix/libc_fatal.c:175
#3  0x00003fff7be27c34 in malloc_printerr (ar_ptr=<optimized out>, ptr=0x3fff700008c0, 
    str=0x3fff7bf2adc0 "double free or corruption (!prev)", action=<optimized out>) at malloc.c:5004
#4  _int_free (have_lock=0, p=<optimized out>, av=<optimized out>) at malloc.c:3865
#5  __GI___libc_free (mem=<optimized out>) at malloc.c:2966
#6  0x00003fff7a384f20 in bshuf_trans_bit_elem_scal (in=<optimized out>, out=0x1000cce7030, size=8192, elem_size=1)
    at src/bitshuffle_core.c:248
#7  0x00003fff7a3855a0 in bshuf_trans_bit_elem (in=<optimized out>, out=<optimized out>, size=<optimized out>, 
    elem_size=<optimized out>) at src/bitshuffle_core.c:1098
#8  0x00003fff7a385668 in bshuf_bitshuffle_block (C_ptr=0x3fffdc7b66a8, size=8192, elem_size=1)
    at src/bitshuffle_core.c:1198
#9  0x00003fff7a384784 in bshuf_blocked_wrap_fun._omp_fn.0 () at src/bitshuffle_core.c:1150
#10 0x00003fff7a320a34 in gomp_thread_start (xdata=<optimized out>) at ../../../libgomp/team.c:119
#11 0x00003fff7c0fbd64 in start_thread (arg=0x3fff776ef190) at pthread_create.c:335
#12 0x00003fff7beba2a0 in .__clone () at ../sysdeps/unix/sysv/linux/powerpc/powerpc64/clone.S:96
```

Running under valgrind:
```
$ valgrind /usr/bin/python3 -m nose tests/
==9290== Memcheck, a memory error detector
==9290== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==9290== Using Valgrind-3.11.0 and LibVEX; rerun with -h for copyright info
==9290== Command: /usr/bin/python3 -m nose tests/
==9290== 
==9290== Thread 3:
==9290== Invalid write of size 1
==9290==    at 0x90F4BF0: bshuf_trans_bit_byte_remainder (bitshuffle_core.c:185)
==9290==    by 0x90F4C93: bshuf_trans_bit_byte_scal (bitshuffle_core.c:197)
==9290==    by 0x90F559F: bshuf_trans_bit_elem (bitshuffle_core.c:1098)
==9290==    by 0x90F5667: bshuf_bitshuffle_block (bitshuffle_core.c:1198)
==9290==    by 0x90F4783: bshuf_blocked_wrap_fun._omp_fn.0 (bitshuffle_core.c:1150)
==9290==    by 0x9150A33: gomp_thread_start (team.c:119)
==9290==    by 0x4CBBD63: start_thread (pthread_create.c:335)
==9290==    by 0x4FAA29F: clone (clone.S:96)
==9290==  Address 0x9256580 is 0 bytes after a block of size 8,192 alloc'd
==9290==    at 0x48962D4: malloc (vg_replace_malloc.c:299)
==9290==    by 0x90F4EAF: bshuf_trans_bit_elem_scal (bitshuffle_core.c:239)
==9290==    by 0x90F559F: bshuf_trans_bit_elem (bitshuffle_core.c:1098)
==9290==    by 0x90F5667: bshuf_bitshuffle_block (bitshuffle_core.c:1198)
==9290==    by 0x90F4783: bshuf_blocked_wrap_fun._omp_fn.0 (bitshuffle_core.c:1150)
==9290==    by 0x9150A33: gomp_thread_start (team.c:119)
==9290==    by 0x4CBBD63: start_thread (pthread_create.c:335)
==9290==    by 0x4FAA29F: clone (clone.S:96)
==9290== 
--9290-- VALGRIND INTERNAL ERROR: Valgrind received a signal 11 (SIGSEGV) - exiting
--9290-- si_code=1;  Faulting address: 0xC0FB92498991B2D0;  sp: 0x80f2afb50

valgrind: the 'impossible' happened:
   Killed by fatal signal
```

--- a/blosc/bitshuffle-generic.c
+++ b/blosc/bitshuffle-generic.c
@@ -10,7 +10,7 @@
 
 
 /* Transpose bytes within elements, starting partway through input. */
-int64_t bshuf_trans_byte_elem_remainder(void* in, void* out, const size_t size,
+int64_t bshuf_trans_byte_elem_remainder(const void* in, void* out, const size_t size,
          const size_t elem_size, const size_t start) {
 
     char* in_b = (char*) in;
@@ -41,7 +41,7 @@ int64_t bshuf_trans_byte_elem_remainder(void* in, void* out, const size_t size,
 
 
 /* Transpose bytes within elements. */
-int64_t bshuf_trans_byte_elem_scal(void* in, void* out, const size_t size,
+int64_t bshuf_trans_byte_elem_scal(const void* in, void* out, const size_t size,
 				   const size_t elem_size) {
 
     return bshuf_trans_byte_elem_remainder(in, out, size, elem_size, 0);
@@ -49,27 +49,35 @@ int64_t bshuf_trans_byte_elem_scal(void* in, void* out, const size_t size,
 
 
 /* Transpose bits within bytes. */
-int64_t bshuf_trans_bit_byte_remainder(void* in, void* out, const size_t size,
+int64_t bshuf_trans_bit_byte_remainder(const void* in, void* out, const size_t size,
          const size_t elem_size, const size_t start_byte) {
 
-    int64_t* in_b = in;
-    int8_t* out_b = out;
+    const uint64_t* in_b = (const uint64_t*) in;
+    uint8_t* out_b = (uint8_t*) out;
 
-    int64_t x, t;
+    uint64_t x, t;
 
+    size_t ii, kk;
     size_t nbyte = elem_size * size;
     size_t nbyte_bitrow = nbyte / 8;
-    size_t ii;
-    int kk;
+
+    uint64_t e=1;
+    const int little_endian = *(uint8_t *) &e == 1;
+    const size_t bit_row_skip = little_endian ? nbyte_bitrow : -nbyte_bitrow;
+    const int64_t bit_row_offset = little_endian ? 0 : 7 * nbyte_bitrow;
 
     CHECK_MULT_EIGHT(nbyte);
     CHECK_MULT_EIGHT(start_byte);
 
     for (ii = start_byte / 8; ii < nbyte_bitrow; ii ++) {
         x = in_b[ii];
-        TRANS_BIT_8X8(x, t);
+        if (little_endian) {
+            TRANS_BIT_8X8(x, t);
+        } else {
+            TRANS_BIT_8X8_BE(x, t);
+        }
         for (kk = 0; kk < 8; kk ++) {
-            out_b[kk * nbyte_bitrow + ii] = x;
+            out_b[bit_row_offset + kk * bit_row_skip + ii] = x;
             x = x >> 8;
         }
     }
@@ -78,7 +86,7 @@ int64_t bshuf_trans_bit_byte_remainder(void* in, void* out, const size_t size,
 
 
 /* Transpose bits within bytes. */
-int64_t bshuf_trans_bit_byte_scal(void* in, void* out, const size_t size,
+int64_t bshuf_trans_bit_byte_scal(const void* in, void* out, const size_t size,
          const size_t elem_size) {
 
     return bshuf_trans_bit_byte_remainder(in, out, size, elem_size, 0);
@@ -86,7 +94,7 @@ int64_t bshuf_trans_bit_byte_scal(void* in, void* out, const size_t size,
 
 
 /* General transpose of an array, optimized for large element sizes. */
-int64_t bshuf_trans_elem(void* in, void* out, const size_t lda,
+int64_t bshuf_trans_elem(const void* in, void* out, const size_t lda,
         const size_t ldb, const size_t elem_size) {
 
     char* in_b = (char*) in;
@@ -103,7 +111,7 @@ int64_t bshuf_trans_elem(void* in, void* out, const size_t lda,
 
 
 /* Transpose rows of shuffled bits (size / 8 bytes) within groups of 8. */
-int64_t bshuf_trans_bitrow_eight(void* in, void* out, const size_t size,
+int64_t bshuf_trans_bitrow_eight(const void* in, void* out, const size_t size,
          const size_t elem_size) {
 
     size_t nbyte_bitrow = size / 8;
@@ -115,7 +123,7 @@ int64_t bshuf_trans_bitrow_eight(void* in, void* out, const size_t size,
 
 
 /* Transpose bits within elements. */
-int64_t bshuf_trans_bit_elem_scal(void* in, void* out, const size_t size,
+int64_t bshuf_trans_bit_elem_scal(const void* in, void* out, const size_t size,
                                   const size_t elem_size, void* tmp_buf) {
 
     int64_t count;
@@ -134,7 +142,7 @@ int64_t bshuf_trans_bit_elem_scal(void* in, void* out, const size_t size,
 
 /* For data organized into a row for each bit (8 * elem_size rows), transpose
  * the bytes. */
-int64_t bshuf_trans_byte_bitrow_scal(void* in, void* out, const size_t size,
+int64_t bshuf_trans_byte_bitrow_scal(const void* in, void* out, const size_t size,
          const size_t elem_size) {
     char* in_b = (char*) in;
     char* out_b = (char*) out;
@@ -157,22 +165,38 @@ int64_t bshuf_trans_byte_bitrow_scal(void* in, void* out, const size_t size,
 
 
 /* Shuffle bits within the bytes of eight element blocks. */
-int64_t bshuf_shuffle_bit_eightelem_scal(void* in, void* out,
+int64_t bshuf_shuffle_bit_eightelem_scal(const void* in, void* out, \
         const size_t size, const size_t elem_size) {
-    char* in_b = (char*) in;
-    char* out_b = (char*) out;
-    size_t nbyte = elem_size * size;
-    int64_t x, t;
-    size_t jj, ii, kk;
+
+    const char *in_b;
+    char *out_b;
+    uint64_t x, t;
+    size_t ii, jj, kk;
+    size_t nbyte, out_index;
+
+    uint64_t e=1;
+    const int little_endian = *(uint8_t *) &e == 1;
+    const size_t elem_skip = little_endian ? elem_size : -elem_size;
+    const uint64_t elem_offset = little_endian ? 0 : 7 * elem_size;
 
     CHECK_MULT_EIGHT(size);
 
+    in_b = (const char*) in;
+    out_b = (char*) out;
+
+    nbyte = elem_size * size;
+
     for (jj = 0; jj < 8 * elem_size; jj += 8) {
         for (ii = 0; ii + 8 * elem_size - 1 < nbyte; ii += 8 * elem_size) {
-            x = *((int64_t*) &in_b[ii + jj]);
-            TRANS_BIT_8X8(x, t);
+            x = *((uint64_t*) &in_b[ii + jj]);
+            if (little_endian) {
+                TRANS_BIT_8X8(x, t);
+            } else {
+                TRANS_BIT_8X8_BE(x, t);
+            }
             for (kk = 0; kk < 8; kk++) {
-                *((uint8_t*) &out_b[ii + jj / 8 + kk * elem_size]) = x;
+                out_index = ii + jj / 8 + elem_offset + kk * elem_skip;
+                *((uint8_t*) &out_b[out_index]) = x;
                 x = x >> 8;
             }
         }
@@ -182,7 +206,7 @@ int64_t bshuf_shuffle_bit_eightelem_scal(void* in, void* out,
 
 
 /* Untranspose bits within elements. */
-int64_t bshuf_untrans_bit_elem_scal(void* in, void* out, const size_t size,
+int64_t bshuf_untrans_bit_elem_scal(const void* in, void* out, const size_t size,
                                     const size_t elem_size, void* tmp_buf) {
 
     int64_t count;

--- a/blosc/bitshuffle-generic.h
+++ b/blosc/bitshuffle-generic.h
@@ -48,6 +48,16 @@ extern "C" {
         x = x ^ t ^ (t << 28);                                              \
     }
 
+/* Transpose 8x8 bit array along the diagonal from upper right
+   to lower left */
+#define TRANS_BIT_8X8_BE(x, t) {                                            \
+        t = (x ^ (x >> 9)) & 0x0055005500550055LL;                          \
+        x = x ^ t ^ (t << 9);                                               \
+        t = (x ^ (x >> 18)) & 0x0000333300003333LL;                         \
+        x = x ^ t ^ (t << 18);                                              \
+        t = (x ^ (x >> 36)) & 0x000000000F0F0F0FLL;                         \
+        x = x ^ t ^ (t << 36);                                              \
+    }
 
 /* Transpose of an array of arbitrarily typed elements. */
 #define TRANS_ELEM_TYPE(in, out, lda, ldb, type_t) {                        \
@@ -72,27 +82,27 @@ extern "C" {
 
 /* Private functions */
 BLOSC_NO_EXPORT int64_t
-bshuf_trans_byte_elem_remainder(void* in, void* out, const size_t size,
+bshuf_trans_byte_elem_remainder(const void* in, void* out, const size_t size,
                                 const size_t elem_size, const size_t start);
 
 BLOSC_NO_EXPORT int64_t
-bshuf_trans_byte_elem_scal(void* in, void* out, const size_t size,
+bshuf_trans_byte_elem_scal(const void* in, void* out, const size_t size,
                            const size_t elem_size);
 
 BLOSC_NO_EXPORT int64_t
-bshuf_trans_bit_byte_remainder(void* in, void* out, const size_t size,
+bshuf_trans_bit_byte_remainder(const void* in, void* out, const size_t size,
                                const size_t elem_size, const size_t start_byte);
 
 BLOSC_NO_EXPORT int64_t
-bshuf_trans_elem(void* in, void* out, const size_t lda,
+bshuf_trans_elem(const void* in, void* out, const size_t lda,
                  const size_t ldb, const size_t elem_size);
 
 BLOSC_NO_EXPORT int64_t
-bshuf_trans_bitrow_eight(void* in, void* out, const size_t size,
+bshuf_trans_bitrow_eight(const void* in, void* out, const size_t size,
                          const size_t elem_size);
 
 BLOSC_NO_EXPORT int64_t
-bshuf_shuffle_bit_eightelem_scal(void* in, void* out,
+bshuf_shuffle_bit_eightelem_scal(const void* in, void* out,
                                  const size_t size, const size_t elem_size);
 
 
@@ -115,7 +125,7 @@ bshuf_shuffle_bit_eightelem_scal(void* in, void* out,
  */
 
 BLOSC_NO_EXPORT int64_t
-bshuf_trans_bit_elem_scal(void* in, void* out, const size_t size,
+bshuf_trans_bit_elem_scal(const void* in, void* out, const size_t size,
                           const size_t elem_size, void* tmp_buf);
 
 /* Unshuffle bitshuffled data.
@@ -140,7 +150,7 @@ bshuf_trans_bit_elem_scal(void* in, void* out, const size_t size,
  */
 
 BLOSC_NO_EXPORT int64_t
-bshuf_untrans_bit_elem_scal(void* in, void* out, const size_t size,
+bshuf_untrans_bit_elem_scal(const void* in, void* out, const size_t size,
                             const size_t elem_size, void* tmp_buf);
 
 

